### PR TITLE
fix: Profile reconcile

### PR DESCRIPTION
**Pull request by genoud**
Use of standard query for EntityDefinition object instead of tooling
Use bulk API when recordCount greater that 2000
Show progress when querying data if log level flag is debug or trace.

Issue #256

### CODE REVIEW

**Comment by genoud** [comment]
@jhawk-kar can you test this on your org and let us know if it fix the issue for you?

**Comment by jhawk-4** [comment]
This is still generating an error for me.  This was the result of running the command with the loglevel set to trace:

```
INFO  [2020-05-05 14:52:32.404 +0000] (sfpowerkit): Reconciling profile ProfileName.profile-meta.xml
DEBUG [2020-05-05 14:52:34.024 +0000] (sfpowerkit): Reconciling App
DEBUG [2020-05-05 14:52:34.026 +0000] (sfpowerkit): Application Visiblitilties reduced from 73  to  72
DEBUG [2020-05-05 14:52:34.027 +0000] (sfpowerkit): Reconciling Classes
DEBUG [2020-05-05 14:52:36.220 +0000] (sfpowerkit): Class Access reduced from 9004  to  8963
DEBUG [2020-05-05 14:52:36.221 +0000] (sfpowerkit): Reconciling Fields
TRACE [2020-05-05 14:52:36.222 +0000] (sfpowerkit): Count Query: SELECT COUNT() FROM ENTITYDEFINITION
TRACE [2020-05-05 14:52:54.761 +0000] (sfpowerkit): Retrieved count 5463
DEBUG [2020-05-05 14:52:54.763 +0000] (sfpowerkit): Using Bulk API
ERROR [2020-05-05 14:53:00.846 +0000] (sfpowerkit): An error occured during profile retrieve. You can rerun the command after a moment.
```

**Comment by genoud** [comment]
> This is still generating an error for me. This was the result of running the command with the loglevel set to trace:
> 
> ```
> INFO  [2020-05-05 14:52:32.404 +0000] (sfpowerkit): Reconciling profile ProfileName.profile-meta.xml
> DEBUG [2020-05-05 14:52:34.024 +0000] (sfpowerkit): Reconciling App
> DEBUG [2020-05-05 14:52:34.026 +0000] (sfpowerkit): Application Visiblitilties reduced from 73  to  72
> DEBUG [2020-05-05 14:52:34.027 +0000] (sfpowerkit): Reconciling Classes
> DEBUG [2020-05-05 14:52:36.220 +0000] (sfpowerkit): Class Access reduced from 9004  to  8963
> DEBUG [2020-05-05 14:52:36.221 +0000] (sfpowerkit): Reconciling Fields
> TRACE [2020-05-05 14:52:36.222 +0000] (sfpowerkit): Count Query: SELECT COUNT() FROM ENTITYDEFINITION
> TRACE [2020-05-05 14:52:54.761 +0000] (sfpowerkit): Retrieved count 5463
> DEBUG [2020-05-05 14:52:54.763 +0000] (sfpowerkit): Using Bulk API
> ERROR [2020-05-05 14:53:00.846 +0000] (sfpowerkit): An error occured during profile retrieve. You can rerun the command after a moment.
> ```

@jhawk-4 can you run again and share the log?

**Comment by jhawk-4** [comment]
```
ERROR [2020-05-05 16:51:34.290 +0000] (sfpowerkit): InvalidBatch : Failed to process query: EXCEEDED_ID_LIMIT: EntityDefinition does not support queryMore(), use LIMIT to restrict the results to a 
single batch
    Error: InvalidBatch : Failed to process query: EXCEEDED_ID_LIMIT: EntityDefinition does not support queryMore(), use LIMIT to restrict the results to a single batch
        at D:\Work\Development\Salesforce\Tools\sfpowerkit\node_modules\jsforce\lib\api\bulk.js:588:32
        at D:\Work\Development\Salesforce\Tools\sfpowerkit\node_modules\jsforce\lib\promise.js:72:9
        at process._tickCallback (internal/process/next_tick.js:61:11)
```

**Comment by genoud** [comment]
Hello @jhawk-4 

We have switch to use describeGlobal to fetch objects as there are some SOQL limitations on EntityDefinition Object as describe here https://developer.salesforce.com/docs/atlas.en-us.api_tooling.meta/api_tooling/reference_objects_soql_limits.htm#topic-title. 

Can you test and let us know if it works on your Org?

Thanks

**Comment by jhawk-4** [comment]
Thanks, @genoud .  I was able to reconcile successfully with the latest changes.  I have a feature request that came to mind out of this that I will open up a separate issue for.  Thanks for all you do to support this effort.

File: src/impl/metadata/retriever/baseMetadataRetriever.ts [hunk 1]
```diff
@@ -1,111 +1,255 @@
...
 
-  private async queryMore(url: string): Promise<QueryResult<T>> {
+  protected async getObjects(): Promise<T[]> {
+    //let records: T[] = [];
     const conn = this.org.getConnection();
-    let result: QueryResult<T>;
+
     if (this.tooling) {
-      result = await conn.tooling.queryMore<T>(url);
+      return executeToolingQueryAsync(this.query, conn, this.objectName);
```
Comment (azlam-abdulsalam) [inline src/impl/metadata/retriever/baseMetadataRetriever.ts:40]: @genoud  await?

File: src/impl/metadata/retriever/baseMetadataRetriever.ts [hunk 2]
```diff
@@ -1,111 +1,255 @@
...
+  let promiseQuery = new Promise<any[]>((resolve, reject) => {
+    let records = [];
+    let hasInitProgress = false;
+    let progressBar = undefined;
+    if (
+      SFPowerkit.logLevel === LoggerLevel.DEBUG ||
+      SFPowerkit.logLevel === LoggerLevel.TRACE
+    ) {
+      progressBar = cli.progress({
+        format: `Querying data from ${object} - PROGRESS  | {bar} | {value}/{total} Records fetched`,
```
Comment (azlam-abdulsalam) [inline src/impl/metadata/retriever/baseMetadataRetriever.ts:78]: @genoud  Can we remove the cli depenedency from the impl layer?

File: src/impl/metadata/retriever/baseMetadataRetriever.ts (hunk 1 above)
Comment (genoud) [reply src/impl/metadata/retriever/baseMetadataRetriever.ts:40]: Yes, Not required as the return type is a promise. The caller can resolve it use await.

File: src/impl/metadata/retriever/baseMetadataRetriever.ts (hunk 2 above)
Comment (genoud) [reply src/impl/metadata/retriever/baseMetadataRetriever.ts:78]: Yes we can do that. May be put it in the SFPowerkit class where we implemented the log.

Comment (azlam-abdulsalam) [reply src/impl/metadata/retriever/baseMetadataRetriever.ts:78]: Sounds good, lets move it over there

Comment (manisfdcsfdx) [reply src/impl/metadata/retriever/baseMetadataRetriever.ts]: @genoud Thanks for the fix.

**Pull request merged**